lib/ruby_marks/default/find_marks.rb

```diff
@@ -0,0 +1,237 @@
+        first_position = 0
+        elements_position_count = 0
+        blocks.map { |m| m[:line] }.each do |dash|
+          marks = blocks.select { |m| m[:line] == dash }
+          if marks.count == group.marks_options.count
+            first_position += marks.first[:x1]
+            elements_position_count += 1
+          end
+        end
+
+        if elements_position_count.positive?
+          first_position /= elements_position_count
+          distance = group.distance_between_marks * (group.marks_options.count - 1)
+          last_position = first_position + distance
+          blocks.delete_if do |mark|
+            mark[:x1] < first_position - 10 ||
+              mark[:x1] > last_position + 10
+          end
+
+          blocks.map { |m| m[:line] }.each do |dash|
+            loop do
+              reprocess = false
+              marks = blocks.select { |m| m[:line] == dash }
+              marks.each_with_index do |current_mark, index|
+                if index.zero?
+                  first_mark_position = first_position - 5..first_position + 5
+                  unless first_mark_position.include?(current_mark[:x1])
+                    new_mark = { x1: first_position,
+                                x2: first_position + group.mark_width,
+                                y1: current_mark[:y1],
+                                y2: current_mark[:y1] + group.mark_height,
+                                line: dash }
+                    blocks << new_mark
+                    blocks.sort_by! { |a| [a[:line], a[:x1]] }
+                    bubbles_adjusted << new_mark
+                    reprocess = true
+                    break
+                  end
+                end
+                next_mark = marks[index + 1]
+                distance = 0
+                distance = next_mark[:x1] - current_mark[:x1] if next_mark
+                next unless distance > group.distance_between_marks + 10 ||
+                            next_mark.nil? && index + 1 < group.marks_options.count
+
+                new_x1 = current_mark[:x1] + group.distance_between_marks
+                new_mark = { x1: new_x1,
+                            x2: new_x1 + group.mark_width,
+                            y1: current_mark[:y1],
+                            y2: current_mark[:y1] + group.mark_height,
+                            line: dash }
+                blocks << new_mark
+                blocks.sort_by! { |a| [a[:line], a[:x1]] }
+                bubbles_adjusted << new_mark
+                reprocess = true
+                break
+              end
+              break unless reprocess
+            end
+          end
+        end
+      end
+
+      def verify_incorrect_bubble_line
+        group.marks.each_pair do |dash, marks|
+          if marks.count != group.marks_options.count
+            incorrect_bubble_line_found[group.label.to_sym] << dash
+          end
+        end
+      end
+
+      attr_reader :group, :image
```

Comment:
I think moving these to the top right after the `private` declaration will improve reliability a little bit.